libs/langchain/src/agents/middlewareAgent/middleware/bigTool.ts

```diff
@@ -0,0 +1,239 @@
+          : runtime.context.maxRetries ?? options.maxRetries;
+      const defaultSystemPrompt =
+        runtime.context.systemPrompt === DEFAULT_SYSTEM_PROMPT
+          ? options.systemPrompt ?? runtime.context.systemPrompt
+          : runtime.context.systemPrompt ?? options.systemPrompt;
+
+      /**
+       * If no tools available, return request unchanged
+       */
+      if (!request.tools || request.tools.length === 0) {
+        return request;
+      }
+
+      /**
+       * Extract tool information
+       */
+      const toolInfo = runtime.tools.map((tool) => ({
+        name: tool.name as string,
+        description: tool.description,
+        tool,
+      }));
+
+      /**
+       * Build tool representation for the prompt
+       */
+      const toolRepresentation = toolInfo
+        .map(({ name, description }) => `- ${name}: ${description}`)
+        .join("\n");
+
+      /**
+       * Build system message
+       */
+      let systemMessage = `You are an agent that can use the following tools:\n${toolRepresentation}\n${defaultSystemPrompt}`;
+
+      if (includeFullHistory) {
+        const userMessages = request.messages
+          .filter(HumanMessage.isInstance)
+          .map((msg: BaseMessage) => msg.content)
+          .join("\n");
+        systemMessage += `\nThe full conversation history is:\n${userMessages}`;
+      }
+
+      if (maxTools !== undefined) {
+        systemMessage += ` You can select up to ${maxTools} tools.`;
+      }
+
+      /**
+       * Get the latest user message
+       */
+      const latestMessage = request.messages.at(-1);
+      const userContent =
+        typeof latestMessage?.content === "string"
+          ? latestMessage?.content
+          : JSON.stringify(latestMessage?.content);
+
+      /**
+       * Create tool selection model
+       */
+      const toolSelectionModel = !model
+        ? (request.model as BaseChatModel)
+        : typeof model === "string"
+        ? await initChatModel(model)
+        : model;
+
+      const validToolNames = toolInfo.map(({ name }) => name);
+      const structuredModel = await toolSelectionModel.withStructuredOutput(
+        ToolSelectionSchema
+      );
+
+      let attempts = 0;
+      let selectedToolNames: string[] = [...validToolNames];
+
+      while (attempts <= maxRetries) {
+        try {
+          const response = await structuredModel.invoke([
+            { role: "system", content: systemMessage },
+            { role: "user", content: userContent },
+          ]);
+
+          selectedToolNames = response.selectedTools;
+
+          /**
+           * Validate that selected tools exist
+           */
+          const invalidTools = selectedToolNames.filter(
+            (name) => !validToolNames.includes(name)
+          );
+
+          if (selectedToolNames.length === 0) {
+            systemMessage += `\n\nNote: You have not selected any tools. Please select at least one tool.`;
```

Comment:
This does not appear to be correct behavior. I think it's OK to select 0 tools if there are no relevant tools.